File: scratch/check_tags.py

import re

file_path = "/Users/taaha/Desktop/Projects/Products/StockFlow/src/pages/OrderDetail.tsx"
with open(file_path, 'r') as f:
    content = f.read()

tokens = re.finditer(r'<(div\b[^>]*?(/)?>)|(</div>)', content, re.DOTALL)

stack = []
for match in tokens:
    tag = match.group(0)
    start_pos = match.start()
    line_num = content.count('\n', 0, start_pos) + 1
    
    if tag.startswith('</'):
        if not stack:
            print(f"[{line_num}] UNEXPECTED CLOSING: {tag}")
        else:
            open_line = stack.pop()
            # print(f"[{line_num}] Closed div from line {open_line}")
    elif tag.endswith('/>'):
        # print(f"[{line_num}] Self-closing: {tag}")
        pass
    else:
        stack.append(line_num)
        # print(f"[{line_num}] Opening: {tag}")

if stack:
    print(f"UNCLOSED TAGS: {stack}")
else:
    print("ALL BALANCED")

File: scratch/check_tags.js

import fs from 'fs';

const content = fs.readFileSync(process.argv[2], 'utf8');
const stack = [];
const regex = /<\/?([a-zA-Z0-9]+)(?:\s+[^>]*?)?>/g;
let match;

while ((match = regex.exec(content)) !== null) {
    const [full, tag] = match;
    const isClosing = full.startsWith('</');
    const isSelfClosing = full.endsWith('/>');

    if (isSelfClosing) continue;

    if (isClosing) {
        if (stack.length === 0) {
            console.log(`Extra closing tag: ${full} at index ${match.index}`);
        } else {
            const last = stack.pop();
            if (last.tag !== tag) {
                console.log(`Mismatched tags: opened ${last.tag} (index ${last.index}), closed ${tag} (index ${match.index})`);
            }
        }
    } else {
        stack.push({ tag, index: match.index, full });
    }
}

while (stack.length > 0) {
    const last = stack.pop();
    console.log(`Unclosed tag: ${last.full} at index ${last.index}`);
}
